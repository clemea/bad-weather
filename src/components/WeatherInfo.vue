<template>
	<div class="weatherInfo">
		<div class="weatherInfo__main">
			<img
				:src="`http://openweathermap.org/img/wn/${weather.weather[0].icon}.png`"
				alt=""
				class="weatherInfo__icon"
			/>
			<div class="weatherInfo__temp">{{ weather.main.temp }} &#176;C</div>
		</div>
		<ul class="weatherInfo__sun">
			<li v-show="sunrise">Рассвет: {{ sunrise }}</li>
			<li v-show="sunset">Закат: {{ sunset }}</li>
		</ul>
		<div v-show="badWeather && !placeNotFound">
			Ugh...
			<span @click="refreshMap" class="weatherInfo__refresher"
				>Take me somewhere!</span
			>
		</div>
		<div v-show="!badWeather && !placeSuggested">Looks good</div>
		<a
			:href="googleMapsLink"
			target="_blank"
			v-if="placeSuggested"
			class="googleLink"
		>
			link
		</a>
		<div v-show="placeNotFound">
			Couldn't find the location that is good enough.
		</div>
	</div>
</template>

<script>
import moment from "moment";
import SunCalc from "suncalc";
import { mapActions, mapState } from "vuex";

export default {
	data() {
		return {};
	},
	computed: {
		...mapState([
			"weather",
			"coords",
			"badWeather",
			"placeSuggested",
			"placeNotFound"
		]),
		sunInfo() {
			if (this.coords) {
				return SunCalc.getTimes(
					new Date(),
					this.coords.lat,
					this.coords.lng
				);
			}
			return null;
		},
		sunrise() {
			return Number.isNaN(Date.parse(this.sunInfo.sunrise))
				? null
				: moment(this.sunInfo.sunrise).format("HH:mm");
		},
		sunset() {
			return Number.isNaN(Date.parse(this.sunInfo.sunset))
				? null
				: moment(this.sunInfo.sunset).format("HH:mm");
		},
		googleMapsLink() {
			return `http://maps.google.com/maps?q=&layer=c&cbll=${
				this.coords.lat
			},${this.coords.lng}&cbp=11,0,0,0,0`;
		}
	},
	methods: {
		...mapActions([
			"setBadWeatherStatus",
			"updateInfo",
			"checkWeather",
			"setPlaceSuggested",
			"getNewPlace"
		]),
		refreshMap() {
			this.getNewPlace();
		}
	}
};
</script>

<style lang="postcss" scoped>
.weatherInfo {
	font-size: 18px;
	padding: 25px;
}
.weatherInfo__temp {
	font-size: 40px;
	margin-bottom: 20px;
}

.weatherInfo__main {
	display: flex;
	align-items: center;
}

.weatherInfo__icon{
	width: 50px;
	flex-shrink: 0;
}
.weatherInfo__sun {
	margin: 0 0 20px;
	padding: 0;
	list-style-type: none;
}

.weatherInfo__refresher {
	font-weight: bold;
	cursor: pointer;
	color: #fff;
	border-bottom: 1px solid #fff;
	transition: all 0.2s ease-in-out;
	&:hover {
		border-bottom-color: transparent;
	}
}

.googleLink {
	color: #fff;
}
</style>
