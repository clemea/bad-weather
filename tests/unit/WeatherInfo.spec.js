import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import WeatherInfo from '@/components/WeatherInfo.vue'
import fakeWeather from '@/__mocks__/weather.json'

const actions = {
	getNewPlace: jest.fn()
}

const state = {
	coords: { lat: 63.23362741232569, lng: -343.84924662625804 },
	weather: fakeWeather,
	badWeather: false,
	placeSuggested: false,
	placeNotFound: false
}

let wrapper
function createWrapper(localstate) {
	const localVue = createLocalVue()
	localVue.use(Vuex)
	const store = new Vuex.Store({
		state: {
			...state,
			...localstate
		},
		actions
	})
	return shallowMount(WeatherInfo, {
		store,
		localVue
	})
}
describe('WeatherInfo.vue', () => {

	test('default template', () => {
		wrapper = createWrapper()
		expect(wrapper.html()).toMatchSnapshot()
	})

	test('Correctly show good weather state', () => {
		wrapper = createWrapper()
		const goodWeather = wrapper.find('.weatherInfo__good')
		const badWeather = wrapper.find('.weatherInfo__bad')
		expect(goodWeather.isVisible()).toBe(true)
		expect(badWeather.isVisible()).toBe(false)
	})

	test('Correctly show bad weather state', () => {
		wrapper = createWrapper({
			badWeather: true
		})
		const goodWeather = wrapper.find('.weatherInfo__good')
		const badWeather = wrapper.find('.weatherInfo__bad')
		expect(goodWeather.isVisible()).toBe(false)
		expect(badWeather.isVisible()).toBe(true)
	})

	test('Correctly show suggested state', () => {
		wrapper = createWrapper({
			placeSuggested: true
		})
		const goodWeather = wrapper.find('.weatherInfo__good')
		const badWeather = wrapper.find('.weatherInfo__bad')
		const googleLink = wrapper.find('.googleLink')
		expect(goodWeather.isVisible()).toBe(false)
		expect(badWeather.isVisible()).toBe(false)
		expect(googleLink.isVisible()).toBe(true)
	})

	test('Show not found msg', () => {
		wrapper = createWrapper({
			placeNotFound: true
		})
		const notFound = wrapper.find('.weatherInfo__notFound')
		expect(notFound.isVisible()).toBe(true)
	})

	test('Show extra link', () => {
		wrapper = createWrapper({
			placeSuggested: true
		})
		const extraLink = wrapper.find('.extraLink')
		expect(extraLink.isVisible()).toBe(true)
	})


	test('Correct google maps link', () => {
		wrapper = createWrapper()
		expect(wrapper.vm.googleMapsLink).toBe('http://maps.google.com/maps?q=&layer=c&cbll=63.23362741232569,-343.84924662625804&cbp=11,0,0,0,0')
	})
})
