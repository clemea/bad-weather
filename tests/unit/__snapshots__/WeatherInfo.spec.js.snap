// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherInfo.vue default template 1`] = `
<div class="weatherInfo">
  <div class="weatherInfo__main"><img src="http://openweathermap.org/img/wn/01d.png" alt="" class="weatherInfo__icon">
    <div class="weatherInfo__temp">18 Â°C</div>
  </div>
  <div class="weatherInfo__bad" style="display: none;">
    Ugh...
    <span class="weatherInfo__refresher">Take me somewhere!</span></div>
  <div class="weatherInfo__good">Looks good</div>
  <!---->
  <div class="extraLink" style="display: none;">
    Another?
  </div>
  <div class="weatherInfo__notFound" style="display: none;">
    Couldn't find the location that is good enough.
  </div>
</div>
`;
